src/script.rs

```diff
@@ -0,0 +1,642 @@
+        assert!(
+            rhs == 0
+                || (rhs > 0 && self.0 <= i64::MAX - rhs)
+                || (rhs < 0 && self.0 >= i64::MIN - rhs)
+        );
+        Self(self.0 + rhs)
+    }
+}
+
+impl Sub for ScriptNum {
+    type Output = Self;
+
+    fn sub(self, other: Self) -> Self {
+        let rhs = other.0;
+        assert!(
+            rhs == 0
+                || (rhs > 0 && self.0 >= i64::MIN + rhs)
+                || (rhs < 0 && self.0 <= i64::MAX + rhs)
+        );
+        Self(self.0 - rhs)
+    }
+}
+
+impl Neg for ScriptNum {
+    type Output = Self;
+
+    fn neg(self) -> Self {
+        assert!(self.0 != i64::MIN);
+        Self(-self.0)
+    }
+}
+
+/** Serialized script, used inside transaction inputs and outputs */
+#[derive(Clone, Debug)]
+pub struct Script<'a>(pub &'a [u8]);
+
+impl Script<'_> {
+    pub fn get_op(script: &mut &[u8]) -> Result<Opcode, ScriptError> {
+        Self::get_op2(script, &mut vec![])
+    }
+
+    pub fn get_op2(script: &mut &[u8], buffer: &mut Vec<u8>) -> Result<Opcode, ScriptError> {
+        if script.is_empty() {
+            return Err(ScriptError::ReadError {
+                expected_bytes: 1,
+                available_bytes: 0,
+            });
+        }
+
+        // Empty the provided buffer, if any
+        buffer.truncate(0);
+
+        let leading_byte = Opcode::from(script[0]);
+        *script = &script[1..];
+
+        Ok(match leading_byte {
+            Opcode::PushValue(pv) => match pv {
+                OP_PUSHDATA1 | OP_PUSHDATA2 | OP_PUSHDATA4 => {
+                    let read_le = |script: &mut &[u8], needed_bytes: usize| {
+                        if script.len() < needed_bytes {
+                            Err(ScriptError::ReadError {
+                                expected_bytes: needed_bytes,
+                                available_bytes: script.len(),
+                            })
+                        } else {
+                            let mut size = 0;
+                            for i in (0..needed_bytes).rev() {
+                                size <<= 8;
+                                size |= usize::from(script[i]);
+                            }
+                            *script = &script[needed_bytes..];
+                            Ok(size)
+                        }
+                    };
+
+                    let size = match pv {
+                        OP_PUSHDATA1 => read_le(script, 1),
+                        OP_PUSHDATA2 => read_le(script, 2),
+                        OP_PUSHDATA4 => read_le(script, 4),
+                        _ => unreachable!(),
+                    }?;
+
+                    if script.len() < size {
+                        return Err(ScriptError::ReadError {
+                            expected_bytes: size,
+                            available_bytes: script.len(),
+                        });
+                    }
+
+                    buffer.extend(&script[0..size]);
```

Comment:
It would be more efficient to just return `Result<(&[u8], Opcode), ScriptError>` (with appropriate lifetime on the slice).